buildSrc/src/main/java/org/elasticsearch/gradle/PublishPlugin.java

```diff
@@ -0,0 +1,202 @@
+                    getArchivesBaseName(project),
+                    project.getVersion()
+                );
+            }
+        }));
+
+        PublishingExtension publishing = project.getExtensions().getByType(PublishingExtension.class);
+
+        project.getPluginManager().withPlugin("com.github.johnrengelman.shadow", plugin -> {
+            MavenPublication publication = publishing.getPublications().maybeCreate("shadow", MavenPublication.class);
+            ShadowExtension shadow = project.getExtensions().getByType(ShadowExtension.class);
+            shadow.component(publication);
+            // Workaround for https://github.com/johnrengelman/shadow/issues/334
+            // Here we manually add any project dependencies in the "shadow" configuration to our generated POM
+            publication.getPom().withXml(xml -> {
+                Node root = xml.asNode();
+                root.appendNode("name", project.getName());
+                root.appendNode("description", project.getDescription());
+                Node dependenciesNode = (Node) ((NodeList) root.get("dependencies")).get(0);
+                project.getConfigurations().getByName(ShadowBasePlugin.getCONFIGURATION_NAME()).getAllDependencies().all(dependency -> {
+                    if (dependency instanceof ProjectDependency) {
+                        Node dependencyNode = dependenciesNode.appendNode("dependency");
+                        dependencyNode.appendNode("groupId", dependency.getGroup());
+                        ProjectDependency projectDependency = (ProjectDependency) dependency;
+                        String artifactId = getArchivesBaseName(projectDependency.getDependencyProject());
+                        dependencyNode.appendNode("artifactId", artifactId);
+                        dependencyNode.appendNode("version", dependency.getVersion());
+                        dependencyNode.appendNode("scope", "compile");
+                    }
+                });
+            });
+        });
+
+        // Add git origin info to generated POM files
+        publishing.getPublications().withType(MavenPublication.class, publication -> {
+            publication.getPom().withXml(PublishPlugin::addScmInfo);
+
+            // have to defer this until archivesBaseName is set
+            project.afterEvaluate(p -> publication.setArtifactId(getArchivesBaseName(project)));
+
+            generatePomTask.configure(
+                t -> t.dependsOn(String.format("generatePomFileFor%sPublication", Util.capitalize(publication.getName())))
+            );
+        });
+
+    }
+
+    private static void addScmInfo(XmlProvider xml) {
+        Node root = xml.asNode();
+        root.appendNode("url", Util.urlFromOrigin(BuildParams.getGitOrigin()));
+        Node scmNode = root.appendNode("scm");
+        scmNode.appendNode("url", BuildParams.getGitOrigin());
+    }
+
+    private static void configureJavadoc(Project project) {
```

Comment:
I think this should better live in the `ElasticsearchJavaPlugin`. This is dealing with the actual generation of javadocs, not creating an artifact to publish. _Every_ java project gets the `javadoc` task created for it, so due to that, I think we should have consistent behavior there, even if we rarely if ever generate javadocs for modules that aren't also published.